packages/beacon-node/src/util/dataColumns.ts

```diff
@@ -1,41 +1,144 @@
-    Math.max(config.CUSTODY_REQUIREMENT, config.NODE_CUSTODY_REQUIREMENT, config.SAMPLES_PER_SLOT)
-  );
-  const custodyMeta = getCustodyColumnsMeta(custodyColumns);
-  return {...custodyMeta, custodyColumns, sampledColumns};
+class CustodyEventEmitter extends (EventEmitter as {new (): StrictEventEmitter<EventEmitter, ICustodyEvents>}) {}
+
+export class CustodyConfig {
+  readonly emitter = new CustodyEventEmitter();
+
+  /**
+   * The number of custody groups the node should subscribe to
+   */
+  private targetCustodyGroupCount: number;
+
+  /**
+   * The number of custody groups the node will advertise to the network
+   */
+  private advertisedCustodyGroupCount: number;
+
+  private config: ChainForkConfig;
+
+  private nodeId: NodeId;
+
+  constructor(nodeId: NodeId, config: ChainForkConfig) {
+    this.config = config;
+    this.nodeId = nodeId;
+    this.targetCustodyGroupCount = Math.max(config.CUSTODY_REQUIREMENT, config.NODE_CUSTODY_REQUIREMENT);
+    this.advertisedCustodyGroupCount = this.targetCustodyGroupCount;
+  }
+
+  getCustodyColumnsWithIndex(): {
+    custodyColumns: ColumnIndex[];
+    custodyColumnsIndex: Uint8Array;
+  } {
+    const custodyColumns = getDataColumns(this.nodeId, this.targetCustodyGroupCount);
+    const custodyColumnsIndex = this.getCustodyColumnsIndex(custodyColumns);
+
+    return {custodyColumns, custodyColumnsIndex};
+  }
+
+  getSampledGroupCount(): number {
+    // TODO: Porting this over to match current behavior, but I think this incorrectly mixes units:
+    // SAMPLES_PER_SLOT is in columns, and CUSTODY_GROUP_COUNT is in groups
+    return Math.max(this.getTargetCustodyGroupCount(), this.config.SAMPLES_PER_SLOT);
+  }
+
+  /**
+   * Data columns sampled by the node as part of custody sampling
+   * https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/das-core.md#custody-sampling
+   *
+   * TODO: Consider race conditions if this updates during sync/backfill
+   */
+  getSampledColumns(): ColumnIndex[] {
+    return getDataColumns(this.nodeId, this.getSampledGroupCount());
+  }
+
+  /**
+   * The number of custody groups the node should subscribe to
+   */
+  getTargetCustodyGroupCount(): number {
+    return this.targetCustodyGroupCount;
+  }
+
+  /**
+   * The number of custody groups the node will advertise to the network
+   */
+  getAdvertisedCustodyGroupCount(): number {
+    return this.advertisedCustodyGroupCount;
+  }
+
+  updateCustodyRequirement(state: CachedBeaconStateAllForks, validatorIndices: ValidatorIndex[]) {
+    const oldSampledGroupCount = this.getSampledGroupCount();
+    const oldAdvertisedGroupCount = this.getAdvertisedCustodyGroupCount();
+
+    this.targetCustodyGroupCount = getValidatorsCustodyRequirement(state, validatorIndices, this.config);
+
+    const newSampledGroupCount = this.getSampledGroupCount();
+
+    if (oldSampledGroupCount !== newSampledGroupCount) {
+      this.emitter.emit(CustodyEvent.samplingGroupCountUpdated, newSampledGroupCount);
```

Comment:
Now that Ive gotten to the last file I see why you used the emitter from the custodyConfig but I still tend to think we should use one of the existing event busses.  I'm wondering if we can remove `custodyConfig` from the `chain` object and isolate it solely to the `network` and then have all consumers reference it from there.  I see a few places that will not be great though.  Need to think about this a bit.   Trying to get the network object passed into the constructor will create an initialization issue in `packages/beacon-node/src/node/nodejs.ts`.    Hmmmm 🤔